### DISCUSSION TITLE
Support for finer granularity of `EventExecutorChooser` management and utilization

### DISCUSSION THREAD

**Discussion by yangty89** [netty/netty#14140] [Ideas]
Hi, I'm working on a [potential improvement](https://github.com/spring-projects/spring-data-redis/issues/2917) for spring-data-redis. While communicating with Redis using Lettuce via `LettuceConnectionFactory`, the component will by default maintain and inject into each `LettuceConnection` instance a **shared connection** that could serve multiple business threads simultaneously. This functionality is implemented based on netty and it enables Lettuce to communicate efficiently with Redis in a [pipelining way](https://github.com/redis/lettuce/wiki/Pipelining-and-command-flushing).

As the `LettuceConnectionFactory` currently creates and utilize only one shared connection, I'm thinking about increasing its quantity, so that we could exploit the parallelism of the underlying machine and gain a better performance. The key point here is that we should **distribute evenly** the shared connections across different IO `EventLoop`s, so as to achieve a "good" parallelism. 

It would be fine if we only use the shared connection to interact with Redis, however, in the case of sending transactional, pipelining or blocking commands to Redis, a `LettuceConnection` would create a new connection called **"dedicated connection"**, which is dedicated (but also tends to be short lived) to the current business thread, in order to process the request in question without risking interfering with other business threads. 

The problem lies in that both the shared connections and the dedicated connections would be bound to the same IO `EventLoopGroup` and thus **share the same** `EventExecutorChooser` instance. In the worst-case scenario, all the shared connections might turn out to be bound to only a half of `EventLoop`s because of the interference of dedicated connections.

I wonder consequently if it's feasible to implement a finer granularity of `EventExecutorChooser` management and utilization. For the problem mentioned above, we might simply address it by introducing another chooser for the shared connections, which functions independently with the original one.

I think maybe we could add a `ChannelOption` for `EventExecutorChooser`, which could be configured by the client, and act accordingly during channel registration. It might look like something below (some code omitted). I would like to know your opinion on this, thanks :)
```java
public abstract class MultithreadEventLoopGroup extends MultithreadEventExecutorGroup implements EventLoopGroup {

    public ChannelFuture register(Channel channel) {
        EventExecutorChooserFactory.EventExecutorChooser chooser = 
                channel.config().getOption(ChannelOption.EVENTEXECUTOR_CHOOSER);
        return Objects.isNull(chooser) ? next().register(channel) : ((EventLoop) chooser.next()).register(channel);
    }

}
```

**Comment by normanmaurer** [comment]
I am a bit confused why you can't just use one `EventLoopGroup` for all and just do the right thing in the `EventExecutorChooser` or just use two different `EventLoopGroup` instances in the first place. Can you clarify a bit why this does not work ?

**Reply by yangty89** [reply]
Thanks for reply :) The reason why the default `EventExecutorChooser` doesn't work here is that it doesn't distinguish between shared connections and dedicated connections while selecting an `EventLoop` for registration, which may lead to an uneven distribution of shared connections.

For example, say we have 4 processors available on the machine, we create an `EventLoopGroup` of 4 `EventLoop`s and want to register 4 shared connections to the `EventLoopGroup`, each to a different `EventLoop` respectively. We maintain those shared connections in `LettuceConnectionFactory`. Whenever the factory's `getConnection` method is called, we verify if there's already enough shared connections created (4 in our example), if not, we create a new **shared connection** and inject it into a newly created `LettuceConnection` instance that will be returned as a `RedisConnection`. Suppose the business thread holding the connection want to communicate with Redis in a transactional, pipelining, or blocking way, `LettuceConnection` will create a **dedicated connection** and use it instead of the shared connection to do the work.

So when the `getConnection` gets first called, we'll create the first shared connection (S1) that is bound to the first `EventLoop` (E1), and the first dedicated connection (D1) that is bound to the second `EventLoop` (E2). If this behavior repeats, we'll have S2 bound to E3, D2 bound to E4, S3 bound to E1, D3 bound to E2, S4 bound to E3 and D4 bound to E4. So instead of being evenly distributed, in this example part of the shared connections end up being bound to the **same `EventLoop`s** (S1 and S3 bound to E1, S2 and S4 bound to E3), which will negatively impact the effect of parallelism.

I've also considered solely implementing a **customized** `EventExecutorChooser` to distinguish between different connections and act accordingly, but it seems hard to achieve as `EventExecutorChooser` only exposes a `next` method that doesn't accept any parameters and has no means to identify the **type of the connection** being registered. However, it would be fine if we can pass a customized chooser during channel bootstrapping (as mentioned above), as we know what kind of the connection we are operating.

Although using two different `EventLoopGroup`s might also be a solution, I think it's relatively heavyweight, as Java platform threads are expensive and we might also suffer more context switches under heavy load compared to using just a single `EventLoopGroup`. I think consequently that supporting for a finer granularity of `EventExecutorChooser` utilization might be a lightweight and better approach, although it would involve some code modifications. Please let me know if it's clearer and makes sense to you, thanks!

**Reply by yangty89** [reply]
Hi @normanmaurer, may I have your opinion on this, or should I provide more details on the topic? Looking forward to your reply, thanks :)

**Comment by normanmaurer** [comment]
@yangty89 let me think a bit more about this...

**Reply by yangty89** [reply]
ok, thanks. Take your time and don't hesitate to contact me if further information is needed.

**Comment by yangty89** [comment]
Hi @normanmaurer, I think that an `EventExecutorChooser` is somewhat like a load balancer, in that they both make decisions to choose a "work node" based on specific strategies and given information. From this standpoint the default behavior of a chooser is similar to that of a load balancer -- they both do the selection in a round-robin manner. And if we want to achieve a "least connections" behavior, we could probably implement a customized chooser which makes decisions based on the number of channels provided by the `Eventloop`'s `registeredChannels` method. In this case, we make use of specific information provided by our "work nodes".

Returning to our topic in question, the chooser relies on the information of a channel (whether it's shared or dedicated) to make an appropriate decision. It's to some extent akin to the "sticky sessions" mode of a load balancer, in that they both make use of specific information provided by connections. From this point of view I think it might be appropriate to expose channels' information to the chooser. Aside from the approach mentioned above, I think that adding to the `EventExecutorChooser` interface a new `next` method that accept an `Object` as its parameter might also be a solution. (In my opinion, it's more graceful but may involves more code modifications) And a customized chooser could be implemented based on it to identify the type of the connections. It may look like something below. I'd like to know your opinion on this, thanks :)

```java
interface EventExecutorChooser {

    /**
     * Returns the new {@link EventExecutor} to use.
     */
    EventExecutor next();

    /**
     * @param obj an object to be used with the returned EventExecutor (e.g. a channel being registered to an EventLoop)
     * @return a selected EventExecutor to use
     */
    default EventExecutor next(Object obj) {return next();} // call the next() method by default to avoid breaking the current API usage
}
```

```java
// add the method below in class MultithreadEventExecutorGroup

    public EventExecutor next(Object object) {
        return chooser.next(object);
    }

// add/modify the methods below in class MultithreadEventLoopGroup

    public EventLoop next(Channel channel) {
        return (EventLoop) super.next(channel);
    }

    public ChannelFuture register(Channel channel) {
        return next(channel).register(channel);
    }
```

```java
// a simple version of customized chooser for identifying the type of the connection and act accordingly
    private static final class CustomizedExecutorChooser implements EventExecutorChooserFactory.EventExecutorChooser {
        private final AtomicInteger sharedConnIdx = new AtomicInteger(); // shared connection index
        private final AtomicInteger connIdx = new AtomicInteger(); // common connection index (e.g. for dedicated connection)
        private final EventExecutor[] executors;

        CustomizedExecutorChooser(EventExecutor[] executors) {
            this.executors = executors;
        }

        @Override
        public EventExecutor next() {
            return executors[connIdx.getAndIncrement() & executors.length - 1]; // use a common index here
        }

        @Override
        public EventExecutor next(Object object) {
            Channel channel = (Channel) object;
            // if it's a shared connection, use the shared connection index for selection, otherwise use a common index
            if (Objects.equals("shared", channel.attr(AttributeKey.valueOf("connectionType")).get())) {
                return executors[sharedConnIdx.getAndIncrement() & executors.length - 1];    
            }
            return executors[connIdx.getAndIncrement() & executors.length - 1];
        }
    }
```
